### ISSUE TITLE
Subject Selection: Using the Prev/Next arrows in banner does not advance subject number in the url

### ISSUE THREAD

**Issue by goplayoutside3** [zooniverse/front-end-monorepo#4977]
## Package
app-project
lib-classifier

## Describe the bug

For projects with sequential subject selection, the subject number is reflected in the url, and a banner appears on the subject image with Prev and Next arrows. When I click on the Prev or Next arrows, the classifier component switches to the prev or next sequential subject, but the url does not change.

## To Reproduce
1. Go to https://www.zooniverse.org/projects/profdrrogerlouismartinez-davila/deciphering-secrets-unlocking-the-manuscripts-of-medieval-spain
2. Click on a workflow and select a subject set
3. Then select a subject and note its number in the url (example: ...subject/79594595)
4. Click on "Next" in the green banner
5. The url will still read ...subject/79594595

## Expected behavior
The url should always reflect the subject number for projects such as Deciphering Secrets.

## Screenshots
![subject 5](https://github.com/zooniverse/front-end-monorepo/assets/23665803/e237b125-e403-477a-86a0-6a0b00372486)

![subject 6](https://github.com/zooniverse/front-end-monorepo/assets/23665803/90a35c1b-67c1-4341-9d48-4c4a856c04a3)

## Additional context

[Banner](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/components/Classifier/components/Banners/components/SubjectSetProgressBanner/SubjectSetProgressBanner.js) component code

Potentially linked to https://github.com/zooniverse/front-end-monorepo/issues/2537

Engaging Crowds: https://github.com/orgs/zooniverse/projects/23

The classifier's `onSubjectChange` callback updates the window location in the NextJS app.
https://github.com/zooniverse/front-end-monorepo/blob/7d19264db4c12b672c8a3520cdc655fbe1abeff6/packages/app-project/src/screens/ClassifyPage/components/ClassifierWrapper/ClassifierWrapper.js#L61-L70

**Comment by eatyourgreens** [comment]
This should have been fixed by #2549.

**Comment by eatyourgreens** [comment]
Works here:
https://www.zooniverse.org/projects/profdrrogerlouismartinez-davila/deciphering-secrets-unlocking-the-manuscripts-of-medieval-spain/classify/workflow/19276/subject-set/106935/subject/79594593

Clicking previous sends me back to:
https://www.zooniverse.org/projects/profdrrogerlouismartinez-davila/deciphering-secrets-unlocking-the-manuscripts-of-medieval-spain/classify/workflow/19276/subject-set/106935/subject/79594592

**Comment by eatyourgreens** [comment]
I’m unable to reproduce this. Tested in mobile Safari on an iPhone.

**Comment by goplayoutside3** [comment]
1. Bug demo in Firefox:

https://github.com/zooniverse/front-end-monorepo/assets/23665803/1242cf20-09c3-44f8-b7c4-fc9183d7947a


2. Bug demo in Chrome:

https://github.com/zooniverse/front-end-monorepo/assets/23665803/a9800e03-ee40-421f-935b-82b411fe2f6e

**Comment by eatyourgreens** [comment]
No problems here in Firefox. The URL updates as expected, and the correct page shows when I reload.

https://github.com/zooniverse/front-end-monorepo/assets/59547/25e15e03-d2dc-4403-846d-0aa8ca9fb9eb

**Comment by eatyourgreens** [comment]
No problems in Edge 114 or Chrome 114 on WIndows 11 either. The subject ID updates in the window location bar, and the page reloads when I refresh.

**Comment by eatyourgreens** [comment]
@goplayoutside3 could this be a bug in the NextJS router?

**Comment by eatyourgreens** [comment]
I think this bug triggers when the classifier is loaded from a URL that doesn't specify a subject.
eg. https://www.zooniverse.org/projects/profdrrogerlouismartinez-davila/deciphering-secrets-unlocking-the-manuscripts-of-medieval-spain/classify/workflow/19276/subject-set/106935

That will fail the `query.subjectID` check here, which is meant to stop this callback from running on projects like PH-TESS and Gravity Spy.
https://github.com/zooniverse/front-end-monorepo/blob/7d19264db4c12b672c8a3520cdc655fbe1abeff6/packages/app-project/src/screens/ClassifyPage/components/ClassifierWrapper/ClassifierWrapper.js#L64-L69

**Comment by eatyourgreens** [comment]
This bug is an example of why you must declare external dependencies for `useEffect`. 

The React docs now recommend avoiding `useEffect` for most common use cases.

**Comment by goplayoutside3** [comment]
This is broken again. 

Go to PRINT https://www.zooniverse.org/projects/printmigrationnetwork/print/classify, pick a subject and note the subject number in the URL. When using the next button or "take me to next available subject" button, the url does not change.

**Comment by eatyourgreens** [comment]
Should #2537 be reopened too?

**Comment by eatyourgreens** [comment]
Might be worth testing whether the router `query` bug is fixed by upgrading to Next 14.

**Comment by goplayoutside3** [comment]
Note it was decided not to fix this Issue in  https://github.com/zooniverse/front-end-monorepo/pull/5776.

Shallowing routing is not supported in Next.js 13, so relying on `router.replace` in reaction to the subject changing in the Classifier component is a blocker for the Next/Prev subject feature going forward. We can't use `history`, or non-shallow routing due to downsides described here: https://github.com/zooniverse/front-end-monorepo/pull/5776#discussion_r1427216645

At the moment, there's a disconnect between wanting the subject id reflected in the url, controlling the url via NextJS, yet controlling the current subject id in the isolated Classifier component. This Issue can only be fixed by a large refactor of the prioritized subject selection architecture, so will remain open until dedicated funding or effort is allocated to that feature.

**Comment by eatyourgreens** [comment]
NB. this bug also means that the page URL doesn’t update when you press Done, or Done & Talk. That makes it difficult for volunteers to comment on Talk, then return to the Classify page and start the next page in an ordered set (see #2537.)

**Comment by eatyourgreens** [comment]
Related Architecture Decision Records:
- [ADR 37: Indexed subject sets](https://github.com/zooniverse/front-end-monorepo/blob/master/docs/arch/adr-37.md). 
- [ADR 38: Next/Previous subjects](https://github.com/zooniverse/front-end-monorepo/blob/master/docs/arch/adr-38.md).

**Comment by eatyourgreens** [comment]
I previously made a start on rewriting the classifier store architecture, to preserve subject state across page visits, in [ADR 42: Record classifier store snapshots](https://github.com/zooniverse/front-end-monorepo/blob/master/docs/arch/adr-42.md). That ADR might provide useful context too.

**Comment by eatyourgreens** [comment]
> Shallowing routing is not supported in Next.js 13…

This is incorrect. Shallow routing is supported in Next 13 and 14: [Shallow routing in the pages router](https://nextjs.org/docs/pages/building-your-application/routing/linking-and-navigating#shallow-routing).
